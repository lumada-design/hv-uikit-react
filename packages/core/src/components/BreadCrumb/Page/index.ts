export { default as pageClasses } from "./pageClasses";
export * from "./pageClasses";
export * from "./Page";
