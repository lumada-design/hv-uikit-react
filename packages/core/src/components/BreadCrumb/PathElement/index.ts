export { default as pathElementClasses } from "./pathElementClasses";
export * from "./pathElementClasses";
export * from "./PathElement";
