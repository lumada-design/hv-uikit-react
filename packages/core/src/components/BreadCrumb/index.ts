export { default as breadCrumbClasses } from "./breadCrumbClasses";
export * from "./breadCrumbClasses";
export * from "./BreadCrumb";
export * from "./PathElement";
export * from "./Page";
