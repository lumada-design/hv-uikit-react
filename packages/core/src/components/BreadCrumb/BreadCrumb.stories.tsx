import { Meta, StoryObj } from "@storybook/react";
import { HvBreadCrumb, HvBreadCrumbProps } from "components";

const data = [
  { label: "Label 1", path: "route1" },
  { label: "Label 2", path: "route2" },
  { label: "Label 3", path: "route3" },
  { label: "Label 4", path: "route4" },
  { label: "Label 5", path: "route5" },
  { label: "Label 6", path: "route6" },
  { label: "Label 7", path: "route7" },
  { label: "Label 8", path: "route8" },
  { label: "Label 9", path: "route9" },
];

const meta: Meta<typeof HvBreadCrumb> = {
  title: "Navigation/Breadcrumb",
  component: HvBreadCrumb,
};
export default meta;

export const Main: StoryObj<HvBreadCrumbProps> = {
  args: {
    maxVisible: 5,
  },
  argTypes: {
    classes: { control: { disable: true } },
    maxVisible: { control: { type: "range", min: 0, max: data.length } },
  },
  render: (args) => {
    return (
      <HvBreadCrumb
        listRoute={data}
        id="breadcrumb1"
        aria-label="Breadcrumb"
        {...args}
      >
        List
      </HvBreadCrumb>
    );
  },
};

export const WithURL: StoryObj<HvBreadCrumbProps> = {
  parameters: {
    docs: {
      description: {
        story: "Breadcrumb sample that generates the paths using an URL",
      },
    },
  },
  render: () => {
    return (
      <HvBreadCrumb
        url="https://hitachivantara.sharepoint.com/sites/DesignSystem/Pattern%20Library/Home.aspx"
        id="breadcrumb4"
        aria-label="Breadcrumb"
      />
    );
  },
};

export const WithURLLimited: StoryObj<HvBreadCrumbProps> = {
  parameters: {
    docs: {
      description: {
        story:
          "Breadcrumb sample that generates the limited paths using an URL.",
      },
    },
  },
  render: () => {
    return (
      <HvBreadCrumb
        url="https://hitachivantara.sharepoint.com/sites/DesignSystem/Pattern%20Library/Home.aspx"
        id="breadcrumb5"
        maxVisible={2}
        aria-label="Breadcrumb"
      />
    );
  },
};

export const WithClickEvents: StoryObj<HvBreadCrumbProps> = {
  parameters: {
    docs: {
      description: {
        story: "Breadcrumb sample that has a onClick defined.",
      },
    },
  },
  render: () => {
    return (
      <HvBreadCrumb
        listRoute={data}
        id="breadcrumb6"
        aria-label="Breadcrumb"
        onClick={(event, elem) => console.log(elem.path)}
      />
    );
  },
};

export const WithLongLabels: StoryObj<HvBreadCrumbProps> = {
  parameters: {
    docs: {
      description: {
        story: "Breadcrumb with long labels.",
      },
    },
  },
  render: () => {
    const longData = [
      { label: "Label 1 with some long text", path: "route1" },
      { label: "Label 2 with some long text", path: "route2" },
      { label: "Label 3 with some long text", path: "route3" },
      { label: "Label 4 with some long text", path: "route4" },
      { label: "Label 5 with some long text", path: "route5" },
    ];

    return (
      <HvBreadCrumb
        listRoute={longData}
        id="breadcrumb7"
        aria-label="Breadcrumb"
      />
    );
  },
};
