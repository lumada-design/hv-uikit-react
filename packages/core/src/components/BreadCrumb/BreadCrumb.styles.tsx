import styled from "@emotion/styled";
import { theme } from "@hitachivantara/uikit-styles";
import { HvTypography } from "components";

export const StyledRoot = styled("nav")({
  display: "flex",
  alignItems: "center",
  zIndex: 0,
});

export const StyledOrderedList = styled("ol")({
  display: "flex",
  paddingLeft: 0,
  marginLeft: `-${theme.space.xs}`,
});

export const StyledTypography = styled((props) => <HvTypography {...props} />)({
  padding: `8px ${theme.space.xs}`,
});
