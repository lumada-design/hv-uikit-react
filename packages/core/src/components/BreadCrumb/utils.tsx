import { MoreOptionsHorizontal } from "@hitachivantara/uikit-react-icons";
import { HvDropDownMenu } from "@core/components";
import { setId } from "@core/utils";

export const removeExtension = (label) =>
  label.includes(".") ? label.substring(0, label.lastIndexOf(".")) : label;

export const pathWithSubMenu = (
  id,
  onClick,
  listRoute,
  maxVisible,
  dropDownMenuProps
) => {
  const nbrElemToSubMenu = listRoute.length - maxVisible;
  const subMenuList = listRoute.slice(1, nbrElemToSubMenu + 1);

  const handleClick = (event, data) => {
    event.preventDefault();

    onClick?.(event, data);
  };

  listRoute.splice(
    1,
    nbrElemToSubMenu,
    <HvDropDownMenu
      id={setId(id, "submenu")}
      icon={<MoreOptionsHorizontal iconSize="S" color="secondary" />}
      dataList={subMenuList}
      {...dropDownMenuProps}
      onClick={onClick != null ? handleClick : undefined}
    />
  );

  return listRoute;
};
