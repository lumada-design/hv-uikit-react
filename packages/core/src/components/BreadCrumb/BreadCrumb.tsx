import { isValidElement, MouseEvent } from "react";

import isNil from "lodash/isNil";

import { useDefaultProps } from "@core/hooks/useDefaultProps";
import { HvBaseProps } from "@core/types/generic";
import { HvDropDownMenuProps } from "@core/components/DropDownMenu";
import { ExtractNames } from "@core/utils/classes";
import { HvTypography } from "@core/components/Typography";

import { HvPathElement } from "./PathElement";
import { HvPage } from "./Page";
import { staticClasses, useClasses } from "./BreadCrumb.styles";
import { pathWithSubMenu, removeExtension } from "./utils";
import { HvBreadCrumbPathElement } from "./types";

export { staticClasses as breadCrumbClasses };

export type HvBreadCrumbClasses = ExtractNames<typeof useClasses>;

export interface HvBreadCrumbProps
  extends HvBaseProps<HTMLDivElement, "onClick"> {
  /** List of breadcrumb. */
  listRoute?: HvBreadCrumbPathElement[];
  /** URL to build the breadcrumb. */
  url?: string;
  /** Number of pages visible. */
  maxVisible?: number;
  /** The component used for the link node. Either a string to use a DOM element or a component. */
  component?: React.ElementType;
  /** Function passed to the component. If defined the component prop is used as the link node. */
  onClick?: (event: MouseEvent<HTMLElement>, data: any) => void;
  /** Props passed down to the DropDownMenu sub-menu component. */
  dropDownMenuProps?: Partial<HvDropDownMenuProps>;
  /** A Jss Object used to override or extend the styles applied to the component. */
  classes?: HvBreadCrumbClasses;
}

/**
 * A breadcrumb is a graphical control element frequently used as a navigational aid.
 */
export const HvBreadCrumb = (props: HvBreadCrumbProps) => {
  const {
    classes: classesProp,
    className,
    id,
    listRoute = [],
    maxVisible,
    url,
    onClick,
    component,
    dropDownMenuProps,
    ...others
  } = useDefaultProps("HvBreadCrumb", props);

  const { classes, cx } = useClasses(classesProp);

  const maxVisibleElem = maxVisible && maxVisible < 2 ? 2 : maxVisible;
  let listPath = listRoute.slice();

  // build the listPath object list
  if (!isNil(url)) {
    listPath = [];

    // get the domain
    const baseUrl = url.match(/^.*\/\/[^/]+/);

    // get url without domain
    const urlWithoutDomain = url.replace(/^.*\/\/[^/]+/, "");

    const pathNames = urlWithoutDomain.split("/").filter((x) => x);

    pathNames.map((elem, index) =>
      listPath.push({
        label: decodeURI(elem),
        path: `${baseUrl}/${pathNames.slice(0, index + 1).join("/")}`,
      })
    );
  }

  const breadcrumbPath =
    maxVisibleElem && listPath.length > maxVisibleElem
      ? pathWithSubMenu(
          id,
          listPath,
          maxVisibleElem,
          onClick,
          dropDownMenuProps
        )
      : listPath;

  return (
    <nav id={id} className={cx(classes.root, className)} {...others}>
      <ol className={classes.orderedList}>
        {listPath.map((elem, index) => {
          const key = `key_${index}`;
          const isLast = index === breadcrumbPath.length - 1;

          return (
            <HvPathElement
              classes={{
                centerContainer: classes.centerContainer,
                separatorContainer: classes.separatorContainer,
              }}
              key={key}
              last={isLast}
            >
              {(isValidElement(elem) && elem) ||
                (isLast && (
                  <HvTypography className={classes.currentPage} variant="body">
                    {removeExtension(elem.label)}
                  </HvTypography>
                )) || (
                  <HvPage
                    elem={elem}
                    classes={{
                      a: classes.a,
                      link: classes.link,
                    }}
                    component={component}
                    onClick={onClick}
                  />
                )}
            </HvPathElement>
          );
        })}
      </ol>
    </nav>
  );
};
