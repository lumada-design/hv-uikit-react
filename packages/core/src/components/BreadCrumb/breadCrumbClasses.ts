import { getClasses } from "@core/utils";

export interface HvBreadCrumbClasses {
  root?: string;
  link?: string;
  orderedList?: string;
  currentPage?: string;
  centerContainer?: string;
  separatorContainer?: string;
  a?: string;
}

const classKeys: string[] = [
  "root",
  "link",
  "orderedList",
  "currentPage",
  "centerContainer",
  "separatorContainer",
  "a",
];

const breadCrumbClasses = getClasses<HvBreadCrumbClasses>(
  classKeys,
  "HvBreadCrumb"
);

export default breadCrumbClasses;
