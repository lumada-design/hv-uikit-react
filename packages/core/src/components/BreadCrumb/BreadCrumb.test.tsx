import { prettyDOM, render } from "@testing-library/react";
import { HvProvider } from "index";
import { describe, expect, it, vi } from "vitest";
import { HvBreadCrumb } from "./BreadCrumb";

const data = [
  { label: "Label 1", path: "route1" },
  { label: "Label 2", path: "route2" },
  { label: "Label 3", path: "route3" },
  { label: "Label 4", path: "route4" },
  { label: "Label 5", path: "route5" },
  { label: "Label 6", path: "route6" },
  { label: "Label 7", path: "route7" },
];

describe("BreadCrumb", () => {
  const { ResizeObserver } = window;

  beforeEach(() => {
    // @ts-ignore
    delete window.ResizeObserver;
    window.ResizeObserver = vi.fn().mockImplementation(() => ({
      observe: vi.fn(),
      unobserve: vi.fn(),
      disconnect: vi.fn(),
    }));
  });

  afterEach(() => {
    window.ResizeObserver = ResizeObserver;
    vi.restoreAllMocks();
  });

  it("should be defined", () => {
    const { container } = render(<HvBreadCrumb />);
    expect(container).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(<HvBreadCrumb listRoute={data} />);
    expect(container).toMatchSnapshot();
  });

  it("should render the correct number of items", () => {
    const { queryAllByRole } = render(<HvBreadCrumb listRoute={data} />);
    expect(queryAllByRole("listitem").length).toBe(7);
  });

  it("should render the correct number of items if maxVisible is specified", () => {
    const { queryAllByRole } = render(
      <HvProvider>
        <HvBreadCrumb listRoute={data} maxVisible={2} />
      </HvProvider>
    );
    expect(queryAllByRole("listitem").length).toBe(3);
  });
});
