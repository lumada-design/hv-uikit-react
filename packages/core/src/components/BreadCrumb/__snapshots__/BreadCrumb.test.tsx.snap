// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BreadCrumb > should render correctly 1`] = `
<div>
  <nav
    class="HvBreadCrumb-root css-b148f9-StyledRoot e17idydo2"
  >
    <ol
      class="HvBreadCrumb-orderedList css-1fqitne-StyledOrderedList e17idydo1"
    >
      <li
        class="HvPathElement-centerContainer HvBreadCrumb-centerContainer css-1sm7i34-StyledLi e14mbcex0"
      >
        <a
          class="HvTypography-root HvTypography-label HvPage-link HvBreadCrumb-link HvPage-label HvPage-a HvBreadCrumb-a HvTypography-noWrap hv-yrsxp3"
          href="route1"
        >
          <div
            class="HvOverflowTooltip-tooltipAnchor css-trvmb4-StyledDataContainer e199gfye0"
          >
            Label 1
          </div>
        </a>
        <div
          class="HvPathElement-separatorContainer HvBreadCrumb-separatorContainer css-f95wjz"
          name="DropRightXS"
        >
          <svg
            focusable="false"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="color0"
              d="m4.05 10.6-.7-.7L7.25 6l-3.9-3.9.7-.7L8.65 6z"
              fill="var(--uikit-colors-secondary_60)"
            />
          </svg>
        </div>
      </li>
      <li
        class="HvPathElement-centerContainer HvBreadCrumb-centerContainer css-1sm7i34-StyledLi e14mbcex0"
      >
        <a
          class="HvTypography-root HvTypography-label HvPage-link HvBreadCrumb-link HvPage-label HvPage-a HvBreadCrumb-a HvTypography-noWrap hv-yrsxp3"
          href="route2"
        >
          <div
            class="HvOverflowTooltip-tooltipAnchor css-trvmb4-StyledDataContainer e199gfye0"
          >
            Label 2
          </div>
        </a>
        <div
          class="HvPathElement-separatorContainer HvBreadCrumb-separatorContainer css-f95wjz"
          name="DropRightXS"
        >
          <svg
            focusable="false"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="color0"
              d="m4.05 10.6-.7-.7L7.25 6l-3.9-3.9.7-.7L8.65 6z"
              fill="var(--uikit-colors-secondary_60)"
            />
          </svg>
        </div>
      </li>
      <li
        class="HvPathElement-centerContainer HvBreadCrumb-centerContainer css-1sm7i34-StyledLi e14mbcex0"
      >
        <a
          class="HvTypography-root HvTypography-label HvPage-link HvBreadCrumb-link HvPage-label HvPage-a HvBreadCrumb-a HvTypography-noWrap hv-yrsxp3"
          href="route3"
        >
          <div
            class="HvOverflowTooltip-tooltipAnchor css-trvmb4-StyledDataContainer e199gfye0"
          >
            Label 3
          </div>
        </a>
        <div
          class="HvPathElement-separatorContainer HvBreadCrumb-separatorContainer css-f95wjz"
          name="DropRightXS"
        >
          <svg
            focusable="false"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="color0"
              d="m4.05 10.6-.7-.7L7.25 6l-3.9-3.9.7-.7L8.65 6z"
              fill="var(--uikit-colors-secondary_60)"
            />
          </svg>
        </div>
      </li>
      <li
        class="HvPathElement-centerContainer HvBreadCrumb-centerContainer css-1sm7i34-StyledLi e14mbcex0"
      >
        <a
          class="HvTypography-root HvTypography-label HvPage-link HvBreadCrumb-link HvPage-label HvPage-a HvBreadCrumb-a HvTypography-noWrap hv-yrsxp3"
          href="route4"
        >
          <div
            class="HvOverflowTooltip-tooltipAnchor css-trvmb4-StyledDataContainer e199gfye0"
          >
            Label 4
          </div>
        </a>
        <div
          class="HvPathElement-separatorContainer HvBreadCrumb-separatorContainer css-f95wjz"
          name="DropRightXS"
        >
          <svg
            focusable="false"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="color0"
              d="m4.05 10.6-.7-.7L7.25 6l-3.9-3.9.7-.7L8.65 6z"
              fill="var(--uikit-colors-secondary_60)"
            />
          </svg>
        </div>
      </li>
      <li
        class="HvPathElement-centerContainer HvBreadCrumb-centerContainer css-1sm7i34-StyledLi e14mbcex0"
      >
        <a
          class="HvTypography-root HvTypography-label HvPage-link HvBreadCrumb-link HvPage-label HvPage-a HvBreadCrumb-a HvTypography-noWrap hv-yrsxp3"
          href="route5"
        >
          <div
            class="HvOverflowTooltip-tooltipAnchor css-trvmb4-StyledDataContainer e199gfye0"
          >
            Label 5
          </div>
        </a>
        <div
          class="HvPathElement-separatorContainer HvBreadCrumb-separatorContainer css-f95wjz"
          name="DropRightXS"
        >
          <svg
            focusable="false"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="color0"
              d="m4.05 10.6-.7-.7L7.25 6l-3.9-3.9.7-.7L8.65 6z"
              fill="var(--uikit-colors-secondary_60)"
            />
          </svg>
        </div>
      </li>
      <li
        class="HvPathElement-centerContainer HvBreadCrumb-centerContainer css-1sm7i34-StyledLi e14mbcex0"
      >
        <a
          class="HvTypography-root HvTypography-label HvPage-link HvBreadCrumb-link HvPage-label HvPage-a HvBreadCrumb-a HvTypography-noWrap hv-yrsxp3"
          href="route6"
        >
          <div
            class="HvOverflowTooltip-tooltipAnchor css-trvmb4-StyledDataContainer e199gfye0"
          >
            Label 6
          </div>
        </a>
        <div
          class="HvPathElement-separatorContainer HvBreadCrumb-separatorContainer css-f95wjz"
          name="DropRightXS"
        >
          <svg
            focusable="false"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="color0"
              d="m4.05 10.6-.7-.7L7.25 6l-3.9-3.9.7-.7L8.65 6z"
              fill="var(--uikit-colors-secondary_60)"
            />
          </svg>
        </div>
      </li>
      <li
        class="HvPathElement-centerContainer HvBreadCrumb-centerContainer css-1sm7i34-StyledLi e14mbcex0"
      >
        <p
          class="HvTypography-root HvTypography-body HvBreadCrumb-currentPage css-6uegx7-StyledTypography e17idydo0 hv-1gxwolr"
        >
          Label 7
        </p>
      </li>
    </ol>
  </nav>
</div>
`;
